erts/emulator/beam/jit/arm/instr_bs.cpp

```diff
@@ -1443,8 +1482,174 @@ struct BscSegment {
 
     Uint error_info;
     Sint effectiveSize;
+    enum action { NONE, ACCUMULATE_FIRST, ACCUMULATE, STORE } action;
 };
 
+static std::vector<BscSegment> bs_combine_segments(
+        const std::vector<BscSegment> segments) {
+    std::vector<BscSegment> segs;
+
+    for (auto seg : segments) {
+        switch (seg.type) {
+        case am_integer: {
+            if (!(0 < seg.effectiveSize && seg.effectiveSize <= 64)) {
+                segs.push_back(seg);
+                continue;
+            }
+
+            if (seg.flags & BSF_LITTLE || segs.size() == 0 ||
+                segs.back().action == BscSegment::NONE) {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+                continue;
+            }
+
+            auto prev = segs.back();
+            if (prev.flags & BSF_LITTLE) {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+                continue;
+            }
+
+            if (prev.effectiveSize + seg.effectiveSize <= 64) {
+                segs.pop_back();
+                prev.effectiveSize += seg.effectiveSize;
+                seg.action = BscSegment::ACCUMULATE;
+                segs.push_back(seg);
+                segs.push_back(prev);
+            } else {
+                seg.action = BscSegment::ACCUMULATE_FIRST;
+                segs.push_back(seg);
+                seg.action = BscSegment::STORE;
+                segs.push_back(seg);
+            }
+            break;
+        }
+        default:
+            segs.push_back(seg);
+            break;
+        }
+    }
+    return segs;
+}
+
+void BeamModuleAssembler::update_bin_state(arm::Gp bin_base,
+                                           arm::Gp bin_offset,
+                                           Sint bit_offset,
+                                           Sint size,
+                                           arm::Gp size_reg) {
+    int cur_bin_offset =
+            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state) +
+            offsetof(struct erl_bits_state, erts_current_bin_);
```

Comment:
Is there any particular reason for splitting this?
```suggestion
    int cur_bin_offset =
            offsetof(ErtsSchedulerRegisters, aux_regs.d.erl_bits_state.erts_current_bin_);
```